### ISSUE TITLE
Can't run 'pub run test' after upgrading to 0.12.3

### ISSUE THREAD

**Issue by dynaxis** [dart-lang/test#133]
I get the error shown in the end of this report, when I run 'pub run test' after upgrading to 0.12.3.
It seems that it works fine in 0.12.2. I'm running Windows 8.1 on a x64 machine.

```
D:\msys64\home\dynaxis\work\xseed>pub run test
Unhandled exception:
Uncaught Error: FileSystemException: Cannot resolve symbolic links, path = 'D:\packages\dart-sdk\bin\dart' (OS Error: 吏
?뺣맂 ?뚯씪??李얠쓣 ???놁뒿?덈떎.
, errno = 2)
Stack Trace:
#0      FileSystemEntity._throwIfError (dart:io/file_system_entity.dart:671)
#1      FileSystemEntity.resolveSymbolicLinksSync (dart:io/file_system_entity.dart:338)
#2      sdkDir.<anonymous closure> (package:test/src/util/io.dart:26)
#3      sdkDir (package:test/src/util/io.dart:28)
#4      sdkDir (package:test/src/util/io.dart:24)
#5      _sdkVersion (package:test/src/util/io.dart:32)
#6      _sdkVersion (package:test/src/util/io.dart:31)
#7      _supportsPubServe (package:test/src/util/io.dart:65)
#8      supportsPubServe (package:test/src/util/io.dart:60)
#9      supportsPubServe (package:test/src/util/io.dart:60)
#10     main.<main_async_body> (package:test/src/executable.dart:109)
#11     Future.Future.microtask.<anonymous closure> (dart:async/future.dart:144)
#12     _asyncRunCallbackLoop (dart:async/schedule_microtask.dart:41)
#13     _asyncRunCallback (dart:async/schedule_microtask.dart:48)
#14     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:96)
#15     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:143)

#0      _rootHandleUncaughtError.<anonymous closure> (dart:async/zone.dart:886)
#1      _asyncRunCallbackLoop (dart:async/schedule_microtask.dart:41)
#2      _asyncRunCallback (dart:async/schedule_microtask.dart:48)
#3      _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:96)
#4      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:143)
```

**Comment by zoechi** [comment]
To me this looks more like a file system or pub cache problem. Have you tried `pub cache repair` and the again `pub get` or `pub upgrade`

**Comment by dynaxis** [comment]
@zanderso per your suggestion, I tried ```pub cache repair``` followed by deletion of pubspec.lock and ```pub get```. It produced the same error. And I even tried trash all of pub cache manually to no avail. I'm using SDK 1.10.1 and downgrading to test@0.12.2 solves the problem.

**Comment by dynaxis** [comment]
FYI, I'm currently using WebStorm. So I have separate folders for dart-sdk and dartium (that is, no DartEditor folder consolidating dart-sdk and dartium folders).

**Comment by dynaxis** [comment]
https://github.com/dart-lang/test/commit/1eb9d79009a215f492ac990abd3751f3c2868e91
I guess the above commit is making difference. The commit coincides with the stacktrace in my reporting.

**Comment by kevmoo** [comment]
@dynaxis that looks ugly – will investigate on our side.

**Comment by nex3** [comment]
It looks like symbolic link resolution is failing on Windows. This may be a `dart:io` bug. @dynaxis, can you translate the error message for us?

**Comment by dynaxis** [comment]
@nex3 Sorry for the messed up error message. Due to some problem in my console setting, it is unreadable to me, too. But according to my test code trying to invoke resolveSymboliclinksSync on non-existing file (that is, ```...\bin\dart``` when there is ```...\bin\dart.exe```), I got something like ```no such file exists```.

As I stated in my last comment, the cause is rather obvious. ```Platform.executable``` returns ```...\bin\dart``` instead of ```...\bin\dart.exe```, and ```resolveSymbolicLinksSync``` requires there exist the given file. That's the cause of this problem.

I hope the canonical way of getting the SDK directory is sorted out soon.

**Comment by kevmoo** [comment]
I

**Comment by kevmoo** [comment]
I'm not able to reproduce this on 1.11.0-dev.3.0

Perhaps this was fixed recently...

**Comment by dynaxis** [comment]
@kevmoo I tried 1.11.0.-dev3.0 this time. And ```Platform.executable``` reports ```...\bin\dart.exe``` now, instead of ```...\bin\dart```. That seems to solve this issue. I guess the release has its implementation of ```Platform.executable``` different from the previous releases.

**Comment by dynaxis** [comment]
The following commit on ```runtime/bin/platform_win.cc``` and related files seems to solve the issue. I'm not so familiar with Win32/64 programming, but ```GetModuleFileName``` seems to return the full path together with the full name of the executable.

https://github.com/dart-lang/sdk/commit/e03ab1743717c39099a9161999e231699098a811

**Comment by nex3** [comment]
@kevmoo We should add a workaround for this so we can continue to support older versions of the SDK.

**Comment by rbishop-bah** [comment]
I see this error too (Win7 x64).  I tried running tests after `pub global activate test 0.12.2` but that failed with the same error.   Please tell me we won't have to wait until SDK 1.11 is released to be able to test again.

**Comment by nex3** [comment]
> I see this error too (Win7 x64).  I tried running tests after `pub global activate test 0.12.2` but that failed with the same error.

The test runner doesn't (and never will) work with `pub global activate`. It needs to be installed in the same package graph as the tests it's running.

> Please tell me we won't have to wait until SDK 1.11 is released to be able to test again.

Certainly not. You can fix this right now by adding an upper bound to your `test` constraint (`test: ">=0.12.0 <0.12.3"`) and running `pub get` to get an older version. I'm also going to put out a fix very soon that adds a workaround.

**Comment by seaneagan** [comment]
@nex3 The changelog says this fix was released in 0.12.3+1, but that version has not yet been published to pub.dartlang.org.

**Comment by nex3** [comment]
Good catch! Tagged and published.
